src/kudu/consensus/routing.cc

```diff
@@ -0,0 +1,519 @@
+                                                  std::move(proxy_topology),
+                                                  std::move(raft_config)));
+  RETURN_NOT_OK(tmp_drt->Flush()); // no lock needed as object is unpublished
+  *drt = std::move(tmp_drt);
+  return Status::OK();
+}
+
+// Read from disk.
+Status DurableRoutingTable::Load(FsManager* fs_manager,
+                                 std::string tablet_id,
+                                 RaftConfigPB raft_config,
+                                 std::shared_ptr<DurableRoutingTable>* drt) {
+  string path = fs_manager->GetProxyMetadataPath(tablet_id);
+
+  ProxyTopologyPB proxy_topology;
+  RETURN_NOT_OK(pb_util::ReadPBContainerFromPath(fs_manager->env(),
+                                                 path,
+                                                 &proxy_topology));
+
+  *drt = std::shared_ptr<DurableRoutingTable>(new DurableRoutingTable(fs_manager,
+                                          std::move(tablet_id),
+                                          std::move(proxy_topology),
+                                          std::move(raft_config)));
+  return Status::OK();
+}
+
+Status DurableRoutingTable::DeleteOnDiskData(FsManager* fs_manager, const string& tablet_id) {
+  string path = fs_manager->GetProxyMetadataPath(tablet_id);
+  RETURN_NOT_OK_PREPEND(fs_manager->env()->DeleteFile(path),
+                        Substitute("Unable to delete durable routing table file for tablet $0",
+                                   tablet_id));
+  return Status::OK();
+}
+
+Status DurableRoutingTable::UpdateProxyTopology(ProxyTopologyPB proxy_topology) {
+  // Take the write lock (does not block readers) and do the slow stuff here.
+  lock_.WriteLock();
+  auto release_write_lock = MakeScopedCleanup([&] { lock_.WriteUnlock(); });
+
+  // Rebuild the routing table.
+  RoutingTable routing_table;
+  if (leader_uuid_) {
+    Status s = routing_table.Init(raft_config_, proxy_topology, *leader_uuid_);
+    if (PREDICT_FALSE(s.IsIncomplete())) {
+      // Log but continue for Incomplete, which is a warning.
+      LOG_WITH_PREFIX(WARNING) << s.ToString();
+    } else {
+      RETURN_NOT_OK(s);
+    }
+  }
+
+  // Only flush the proxy graph protobuf to disk when it changes.
+  if (!MessageDifferencer::Equals(proxy_topology, proxy_topology_)) {
+    VLOG_WITH_PREFIX(3) << "proxy routes updated, flushing to disk...";
+    RETURN_NOT_OK(Flush());
+  }
+
+  // Upgrade to an exclusive commit lock and make atomic changes here.
+  lock_.UpgradeToCommitLock();
+  release_write_lock.cancel(); // Unlocking the commit lock releases the write lock.
+  auto release_commit_lock = MakeScopedCleanup([&] { lock_.CommitUnlock(); });
+
+  proxy_topology_ = std::move(proxy_topology);
+
+  if (leader_uuid_) {
+    routing_table_ = std::move(routing_table);
+    VLOG_WITH_PREFIX(2) << "updated proxy routes: \n" << routing_table_->ToString();
+  } else {
+    routing_table_ = boost::none;
```

Comment:
I am not sure I understand this design. Requirement of it being a routing tree, instead of just a 
next hop table might be the reason.
When leader switches the routing table of most  nodes should not have changed.

Reply:
The way it's designed, the tree is constructed by making all nodes without parents children of the leader. Without the leader it remains a forest and not all nodes are connected. When we don't know what to do we don't construct the RoutingTable and instead directly route wherever possible. However in practice this is mostly bookkeeping and we should generally know who the leader is because when routing a request that request is coming from the leader.